src/include/homestore/blk.h

```diff
@@ -24,122 +24,162 @@
-    // last blk num is the last blk num that belongs this blkid;
-    blk_num_t get_last_blk_num() const { return get_blk_num() + get_nblks() - 1; }
+#pragma pack(1)
+struct MultiBlkId : public BlkId {
+    static constexpr uint32_t max_addln_pieces{5};
+    static constexpr uint32_t max_pieces{max_addln_pieces + 1};
 
-    void set_nblks(blk_count_t nblks);
-    blk_count_t get_nblks() const { return static_cast< blk_count_t >(m_nblks) + 1; }
+private:
+    struct chain_blkid {
+        blk_num_t m_blk_num;
+        blk_count_t m_nblks{0};
 
-    void set_chunk_num(const chunk_num_t chunk_num);
-    chunk_num_t get_chunk_num() const { return m_chunk_num; }
+        bool is_valid() const { return (m_nblks != 0); }
+    };
 
-    /* A blkID represent a page size which is assigned to a blk allocator */
-    uint32_t data_size(const uint32_t page_size) const { return (get_nblks() * page_size); }
+    uint16_t n_addln_piece{0};
+    std::array< chain_blkid, max_addln_pieces > addln_pieces;
 
+public:
+    MultiBlkId();
+    MultiBlkId(BlkId const& b);
+    MultiBlkId(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num);
+    MultiBlkId(MultiBlkId const&) = default;
+    MultiBlkId& operator=(MultiBlkId const&) = default;
+    MultiBlkId(MultiBlkId&&) noexcept = default;
+    MultiBlkId& operator=(MultiBlkId&&) noexcept = default;
+
+    void add(blk_num_t blk_num, blk_count_t nblks, chunk_num_t chunk_num);
+    void add(BlkId const&);
+
+    uint16_t num_pieces() const;
+    blk_count_t blk_count() const;
     std::string to_string() const;
 
-    blk_num_t m_blk_num;            // Block number which is unique within the chunk
-    blk_count_serialized_t m_nblks; // Number of blocks+1 for this blkid, don't directly acccess this - use get_nblks()
-    chunk_num_t m_chunk_num;        // Chunk number - which is unique for the entire application
-} __attribute__((__packed__));
-
-VENUM(BlkAllocStatus, uint32_t,
-      BLK_ALLOC_NONE = 0,       // No Action taken
-      SUCCESS = 1ul << 0,       // Success
-      FAILED = 1ul << 1,        // Failed to alloc/free
-      REQ_MORE = 1ul << 2,      // Indicate that we need more
-      SPACE_FULL = 1ul << 3,    // Space is full
-      INVALID_DEV = 1ul << 4,   // Invalid Device provided for alloc
-      PARTIAL = 1ul << 5,       // In case of multiple blks, only partial is alloced/freed
-      INVALID_THREAD = 1ul << 6 // Not possible to alloc in this thread
-);
-
-static_assert(sizeof(BlkId) < 8);
-#pragma pack(1)
-struct BlkId8_t : public BlkId {
-    uint8_t pad[8 - sizeof(BlkId)]{};
-
-    BlkId8_t& operator=(const BlkId& rhs) {
-        BlkId::operator=(rhs);
-        return *this;
-    }
+    bool operator==(MultiBlkId const& other) const { return (compare(*this, other) == 0); }
+    bool operator>(MultiBlkId const& other) const { return (compare(*this, other) > 0); }
+    bool operator<(MultiBlkId const& other) const { return (compare(*this, other) < 0); }
+
+    sisl::blob serialize();
```

Comment:
I can do const for sure, can't do override I think since its not a virtual derived class.

Reply:
Since it is directly returning uint8_t*, const will fail. Thats why I remember removing const from this. I have added a TODO to handle this, by doing something like sisl::const_blob for reasons like these.